go/kbfs/libkbfs/subscription_manager.go

```diff
@@ -0,0 +1,331 @@
+package libkbfs
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/data"
+	"github.com/keybase/client/go/kbfs/tlfhandle"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+)
+
+// SubscriptionID identifies a subscription.
+type SubscriptionID string
+
+// SubscriptionNotifier defines a group of methods for notifying about changes
+// on subscribed topics.
+type SubscriptionNotifier interface {
+	OnPathChange(subscriptionID SubscriptionID, path string, topic keybase1.PathSubscriptionTopic)
+	OnChange(subscriptionID SubscriptionID, topic keybase1.SubscriptionTopic)
+}
+
+// Subscriber defines a type that can be used to subscribe to different topic.
+type Subscriber interface {
+	SubscribePath(
+		ctx context.Context, path string, topic keybase1.PathSubscriptionTopic,
+		deduplicateInterval *time.Duration) (SubscriptionID, error)
+	SubscribeNonPath(ctx context.Context, topic keybase1.SubscriptionTopic,
+		deduplicateInterval *time.Duration) (SubscriptionID, error)
+	Unsubscribe(context.Context, SubscriptionID)
+}
+
+// SubscriptionManager manages subscriptions. Use the Subscriber interface to
+// subscribe and unsubscribe. Multiple subscribers can be used with the same
+// SubscriptionManager.
+type SubscriptionManager interface {
+	Subscriber(SubscriptionNotifier) Subscriber
+	Shutdown(ctx context.Context)
+}
+
+// SubscriptionManagerPublisher associates with one SubscriptionManager, and is
+// used to publish changes to subscribers mangaged by it.
+type SubscriptionManagerPublisher interface {
+	FavoritesChanged()
+	JournalStatusChanged()
+}
+
+// userPath is always the full path including the /keybase prefix, but may
+// not be canonical or normalized. For examplef, it can be
+// . The goal is to track
+// whatever the user of this type is dealing with without needing them to
+// know if a path is canonicalized at any time.
+// Examples:
+//   "/keybase/public/karlthefog@twitter/dir
+//   "/keybase/team/keybase/dir/../file"
+type userPath string
+
+// normalizedInTlfPath is normalized path rooted at a TLF, and it's what we get
+// from Node.GetPathPlaintextSansTlf().
+// Examples, considering TLF /keybase/private/user1,user2:
+//   "/foo/bar" (representing /keybsae/private/user1,user2/foo/bar)
+//   "/"        (representing /keybsae/private/user1,user2)
+type normalizedInTlfPath string
+
+type debouncedNotify struct {
+	notify   func()
+	shutdown func()
+}
+
+func debounce(do func(), limit rate.Limit) debouncedNotify {
+	ctx, shutdown := context.WithCancel(context.Background())
+	ch := make(chan struct{})
+	limiter := rate.NewLimiter(limit, 1)
+	go func() {
+		for {
+			limiter.Wait(ctx)
+			select {
+			case <-ch:
+				go do()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return debouncedNotify{
+		notify: func() {
+			select {
+			case ch <- struct{}{}:
+			default:
```

Comment:
If deduplicateInterval is set, we do want to drop notifications. Though I guess when it's not set this can potentially drop as well if the other side of the channel doesn't spawn goroutines fast enough. I don't think making it buffered is a best answer though since it'd buffer one more when debounced. I'll make it a blocking send when deduplicateInterval is not set, but let me know if you think that's a bad idea!